guardian/admin.py

```diff
@@ -428,6 +428,66 @@ class AuthorAdmin(GuardedModelAdmin):
         admin.site.register(Author, AuthorAdmin)
 
     """
+    def get_model_objs(self, request, action=None, klass=None):
+        opts = self.opts
+        actions = [action] if action else ['view', 'change', 'delete']
+        klass = klass if klass else opts.model
+        model_name = klass._meta.model_name
+        return get_objects_for_user(user=request.user, perms=['{}_{}'.format(perm, model_name) for perm in actions],
+                                    klass=klass, any_perm=True)
+
+    def has_module_permission(self, request):
+        if super(GuardedModelAdmin, self).has_module_permission(request):
+            return True
+        return self.get_model_objs(request).exists()
+
+    def get_queryset(self, request):
+        if request.user.is_superuser:
+            return super(GuardedModelAdmin, self).get_queryset(request)
+
+        data = self.get_model_objs(request)
+        return data
+
+    def has_perm(self, request, obj, action):
+        opts = self.opts
+        codename = '{action}_{model_name}'.format(action=action, model_name=opts.model_name)
+        if obj:
+            return request.user.has_perm('{label}.{codename}'.format(label=opts.app_label, codename=codename), obj)
+        else:
+            return self.get_model_objs(request, action).exists()
+
+    def has_view_permission(self, request, obj=None):
+        return self.has_perm(request, obj, 'view')
+
+    def has_change_permission(self, request, obj=None):
+        return self.has_perm(request, obj, 'change')
+
+    def has_delete_permission(self, request, obj=None):
+        return self.has_perm(request, obj, 'delete')
+
+    def save_model(self, request, obj, form, change):
+        result = super(GuardedModelAdmin, self).save_model(request, obj, form, change)
+        if not request.user.is_superuser and not change:
+            opts = self.opts
+            actions = ['view', 'add', 'change', 'delete']
+            [assign_perm('{}.{}_{}'.format(opts.app_label, action, opts.model_name), request.user, obj)
+             for action in actions]
+        return result
+
+    @staticmethod
+    def remove_obj_perms(obj):
+        perms_dict = get_users_with_perms(obj, attach_perms=True)
+        perms_dict.update(get_groups_with_perms(obj, attach_perms=True))
+        for user_or_group in perms_dict:
+            [remove_perm(perm, user_or_group, obj) for perm in perms_dict[user_or_group]]
```

Comment:
We extract the code for "shortcuts" instead of static methods. This allows you to re-use the code.

You download the permissions (+1 query) to then delete the permission for each user (+n queries). Could you eliminate this and build the correct query that will do it all at once? Maybe we need to extend our sho

Reply:
When an obj is deleted, we need have an api to delete its permission. I think this api put to "shortcuts" is good.

Reply:
Could you update pull request to reflect that?